File: src/storage.js
import { copyFile, mkdir } from "node:fs/promises";
import path from "node:path";
import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { readFile } from "node:fs/promises";

export async function persistScreenshots(config, files) {
  const localOutputDir = process.env.LOCAL_OUTPUT_DIR;
  if (localOutputDir) {
    return saveLocally(config, files, localOutputDir);
  }

  if (!config.s3Bucket) {
    throw new Error(
      "S3_BUCKET is not configured. Set LOCAL_OUTPUT_DIR for local file output.",
    );
  }

  return uploadToS3(config, files);
}

async function saveLocally(config, files, outputDir) {
  const prefixDir = path.join(
    outputDir,
    config.s3Prefix.replace(/^\/+|\/+$/g, ""),
  );
  await mkdir(prefixDir, { recursive: true });
  const publicUrls = [];

  for (const filePath of files) {
    const destination = path.join(prefixDir, path.basename(filePath));
    await copyFile(filePath, destination);
    const publicUrl = `${config.publicBaseUrl}/${path.basename(filePath)}`;
    publicUrls.push(publicUrl);
    console.log(`Saved locally: ${destination}`);
  }

  return publicUrls;
}

async function uploadToS3(config, files) {
  const client = new S3Client({});
  const publicUrls = [];

  for (const filePath of files) {
    const key = `${config.s3Prefix}${path.basename(filePath)}`;
    const body = await readFile(filePath);
    await client.send(
      new PutObjectCommand({
        Bucket: config.s3Bucket,
        Key: key,
        Body: body,
        ContentType: "image/jpeg",
      }),
    );
    const publicUrl = `${config.publicBaseUrl}/${path.basename(filePath)}`;
    publicUrls.push(publicUrl);
    console.log(`Uploaded s3://${config.s3Bucket}/${key}`);
  }

  return publicUrls;
}

File: src/storage.py
from __future__ import annotations

import logging
import mimetypes
import os
import shutil
from pathlib import Path

import boto3

from config import AppConfig

logger = logging.getLogger(__name__)


def persist_screenshots(config: AppConfig, files: list[Path]) -> list[str]:
    local_output_dir = os.environ.get("LOCAL_OUTPUT_DIR")
    if local_output_dir:
        return _save_locally(config, files, Path(local_output_dir))

    if not config.s3_bucket:
        raise ValueError(
            "S3_BUCKET is not configured. Set LOCAL_OUTPUT_DIR for local file output."
        )

    return _upload_to_s3(config, files)


def _save_locally(config: AppConfig, files: list[Path], output_dir: Path) -> list[str]:
    prefix_dir = output_dir / config.s3_prefix.strip("/")
    prefix_dir.mkdir(parents=True, exist_ok=True)
    public_urls: list[str] = []

    for file_path in files:
        destination = prefix_dir / file_path.name
        shutil.copy2(file_path, destination)
        public_url = f"{config.public_base_url}/{file_path.name}"
        public_urls.append(public_url)
        logger.info("Saved locally: %s", destination)

    return public_urls


def _upload_to_s3(config: AppConfig, files: list[Path]) -> list[str]:
    client = boto3.client("s3")
    public_urls: list[str] = []

    for file_path in files:
        key = f"{config.s3_prefix}{file_path.name}"
        content_type = mimetypes.guess_type(file_path.name)[0] or "image/jpeg"
        client.upload_file(
            str(file_path),
            config.s3_bucket,
            key,
            ExtraArgs={"ContentType": content_type},
        )
        public_url = f"{config.public_base_url}/{file_path.name}"
        public_urls.append(public_url)
        logger.info("Uploaded s3://%s/%s", config.s3_bucket, key)

    return public_urls
